File: algorithms/java/src/test/java/leetcode/LC328OddEvenLinkedList.java
package leetcode;

import org.junit.Test;

import java.util.List;

import static org.junit.Assert.assertEquals;

/**
 * https://leetcode.com/problems/odd-even-linked-list/
 * #medium
 */
public class LC328OddEvenLinkedList {
    public class ListNode {
        int val;
        ListNode next;

        ListNode(int x) {
            val = x;
        }
    }

    public class Solution {
        public ListNode oddEvenList(ListNode head) {
            ListNode evenHead = new ListNode(0);
            ListNode oddHead = new ListNode(0);
            ListNode evenTail = evenHead;
            ListNode oddTail = oddHead;
            boolean odd = true;
            while (head != null) {
                if (odd) {
                    oddTail.next = head;
                    oddTail = oddTail.next;
                    odd = false;
                } else {
                    evenTail.next = head;
                    evenTail = evenTail.next;
                    odd = true;
                }
                head = head.next;
            }
            evenTail.next = null;
            oddTail.next = evenHead.next;
            return oddHead.next;
        }
    }

    @Test
    public void test_example() throws Exception {
        ListNode head = new ListNode(0);
        ListNode tail = head;
        for (int i = 1; i <= 5; i++) {
            tail.next = new ListNode(i);
            tail = tail.next;
        }
        ListNode result = new Solution().oddEvenList(head.next);
        int[] expected = {1, 3, 5, 2, 4};
        for (int value : expected) {
            assertEquals(value, result.val);
            result = result.next;
        }
        assertEquals(null, result);
    }

    @Test
    public void test_null() throws Exception {
        assertEquals(null, new Solution().oddEvenList(null));
    }
}

File: algorithms/code/leetcode/lc328_odd_even_linked_list/lc328_odd_even_linked_list.rs
// https://leetcode.com/problems/odd-even-linked-list/
// #medium

#[derive(PartialEq, Eq, Clone, Debug)]
pub struct ListNode {
    pub val: i32,
    pub next: Option<Box<ListNode>>,
}

pub struct Solution;

impl Solution {
    pub fn odd_even_list(head: Option<Box<ListNode>>) -> Option<Box<ListNode>> {
        let mut odd_head = Box::new(ListNode { val: 0, next: None });
        let mut even_head = Box::new(ListNode { val: 0, next: None });
        let mut odd_tail = &mut odd_head;
        let mut even_tail = &mut even_head;
        let mut head = head;
        let mut odd = true;
        while let Some(mut some_head) = head {
            head = some_head.next.take();
            if odd {
                odd_tail.next = Some(some_head);
                odd_tail = odd_tail.next.as_mut().unwrap();
                odd = false;
            } else {
                even_tail.next = Some(some_head);
                even_tail = even_tail.next.as_mut().unwrap();
                odd = true;
            }
        }
        odd_tail.next = even_head.next.take();
        odd_head.next
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    fn vector_to_list(vector: Vec<i32>) -> Option<Box<ListNode>> {
        let mut list: Option<Box<ListNode>> = None;
        for &value in vector.iter().rev() {
            let node = Box::new(ListNode {
                val: value,
                next: list,
            });
            list = Some(node);
        }
        list
    }

    fn list_to_vector(mut list: Option<Box<ListNode>>) -> Vec<i32> {
        let mut vector = Vec::new();
        while let Some(node) = list {
            vector.push(node.val);
            list = node.next;
        }
        vector
    }

    #[test]
    fn test_example_1() {
        let list = vector_to_list(vec![1, 2, 3, 4, 5]);
        let expected = vec![1, 3, 5, 2, 4];
        let result = Solution::odd_even_list(list);
        assert_eq!(list_to_vector(result), expected);
    }

    #[test]
    fn test_nothing() {
        let result = Solution::odd_even_list(None);
        assert!(result.is_none());
    }
}
